File: UI/clinicnow_mobile/lib/core/contact_rules.dart
/// Client-side mirror of `ClinicNow.Services/Validation/ContactRules.cs`
/// (review item C17).
///
/// The backend is the authority - it re-validates every one of these rules and
/// nothing here is trusted (rulebook §5). These exist so the user is told what
/// is wrong *under the field* as they type, instead of round-tripping to the
/// API for a message they could have had immediately (rulebook §4).
///
/// The patterns, limits and messages are deliberately identical to the C# ones.
/// If you change a rule, change it in both places - a client rule that is
/// stricter than the server's silently blocks legitimate input, and one that is
/// looser just moves the error to the submit button.
class ContactRules {
  ContactRules._();

  // Mirrors the EF column limits the values land in.
  static const int maxEmailLength = 256;
  static const int maxPhoneLength = 30;
  static const int maxNameLength = 100;
  static const int maxPersonalIdLength = 20;
  static const int maxAddressLength = 250;
  static const int maxLicenseNumberLength = 50;
  static const int maxBioLength = 1000;
  static const int minPasswordLength = 8;

  static const String emailMessage =
      'Unesite ispravnu email adresu (npr. ime@primjer.com).';
  static const String phoneMessage =
      'Unesite ispravan broj telefona (npr. +38761123456).';

  static final RegExp _emailPattern = RegExp(r'^[^@\s]+@[^@\s]+\.[^@\s]+$');
  static final RegExp _phonePattern = RegExp(r'^\+?[0-9 ]{6,20}$');

  /// Returns null when valid, or the message to render under the field.
  static String? email(String? value, {bool required = false}) {
    final trimmed = value?.trim() ?? '';
    if (trimmed.isEmpty) {
      return required ? emailMessage : null;
    }
    if (!_emailPattern.hasMatch(trimmed)) return emailMessage;
    if (trimmed.length > maxEmailLength) {
      return 'Email adresa može imati najviše $maxEmailLength karaktera.';
    }
    return null;
  }

  /// Phone is optional everywhere in this app, but a supplied one must be real.
  static String? phone(String? value) {
    final trimmed = value?.trim() ?? '';
    if (trimmed.isEmpty) return null;
    if (!_phonePattern.hasMatch(trimmed)) return phoneMessage;
    if (trimmed.length > maxPhoneLength) {
      return 'Broj telefona može imati najviše $maxPhoneLength karaktera.';
    }
    return null;
  }

  /// Free text held to its column length, and optionally to being present.
  static String? text(
    String? value, {
    required int maxLength,
    required String label,
    bool required = false,
  }) {
    final trimmed = value?.trim() ?? '';
    if (trimmed.isEmpty) {
      return required ? '$label je obavezno.' : null;
    }
    if (trimmed.length > maxLength) {
      return '$label može imati najviše $maxLength karaktera.';
    }
    return null;
  }
}

File: ClinicNow.Services/Validation/ContactRules.cs
using System.Text.RegularExpressions;

namespace ClinicNow.Services.Validation;

/// <summary>
/// The single definition of what a valid email, phone number and free-text
/// length looks like (review item C17). Every flow that fills the same column
/// asks the same question here, so registration, staff-created patients and
/// admin-created doctors can no longer disagree about what they accept.
///
/// Before this, they did: registration matched a real email pattern and a phone
/// pattern, <c>PatientService</c> accepted anything containing an "@" and never
/// looked at the phone at all, and <c>DoctorService</c> checked email
/// *uniqueness* with no format check whatsoever - so the same address was
/// rejected on one screen and stored on another.
///
/// The length limits mirror the EF configuration for the columns these values
/// land in. Without them an over-long value passes service validation and then
/// fails deep in SQL as a truncation error, which reaches the user as an
/// unexplained 500 instead of a message under the field (rulebook §4).
///
/// Same reasoning as <see cref="ClinicNow.Services.Documents.FileValidation"/>
/// and <see cref="ClinicNow.Services.Appointments.DoctorCompatibility"/>: one
/// rule, many callers, never a second copy.
/// </summary>
public static partial class ContactRules
{
    /// <summary>
    /// Matches <c>UserConfiguration</c>'s <c>Email</c> column. <c>Patient.Email</c>
    /// is configured wider (320), but a patient's address may later become a
    /// login, so the stricter of the two is the shared bound - anything that
    /// passes here fits both columns.
    /// </summary>
    public const int MaxEmailLength = 256;

    public const int MaxPhoneLength = 30;   // User.PhoneNumber / Patient.PhoneNumber
    public const int MaxNameLength = 100;   // First/LastName on both User and Patient
    public const int MaxPersonalIdLength = 20;
    public const int MaxAddressLength = 250;
    public const int MaxLicenseNumberLength = 50;  // Doctor.LicenseNumber
    public const int MaxBioLength = 1000;          // Doctor.Bio
    public const int MinPasswordLength = 8;

    // Messages name the expected format rather than just saying "invalid"
    // (rulebook §4: the message states the constraint).
    public const string EmailMessage = "Unesite ispravnu email adresu (npr. ime@primjer.com).";
    public const string PhoneMessage = "Unesite ispravan broj telefona (npr. +38761123456).";

    /// <summary>Required email: must be present, well-formed and within the column length.</summary>
    public static void RequireEmail(IDictionary<string, string[]> errors, string field, string? value)
    {
        if (string.IsNullOrWhiteSpace(value) || !EmailPattern().IsMatch(value.Trim()))
        {
            errors[field] = [EmailMessage];
            return;
        }

        CheckLength(errors, field, value, MaxEmailLength, "Email adresa");
    }

    /// <summary>Optional email: blank is fine, but anything supplied is held to the same rule.</summary>
    public static void OptionalEmail(IDictionary<string, string[]> errors, string field, string? value)
    {
        if (!string.IsNullOrWhiteSpace(value))
        {
            RequireEmail(errors, field, value);
        }
    }

    /// <summary>Optional phone - no flow here makes one mandatory, but a supplied number must be real.</summary>
    public static void OptionalPhone(IDictionary<string, string[]> errors, string field, string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return;
        }

        if (!PhonePattern().IsMatch(value.Trim()))
        {
            errors[field] = [PhoneMessage];
            return;
        }

        CheckLength(errors, field, value, MaxPhoneLength, "Broj telefona");
    }

    /// <summary>
    /// Required free text (a name): present, and short enough for its column.
    /// The "je obavezno" phrasing matches the labels that actually use this
    /// today (Ime, Prezime - both neuter) and the wording the Flutter forms
    /// already show; a feminine label would need its own message rather than
    /// this template.
    /// </summary>
    public static void RequireText(
        IDictionary<string, string[]> errors, string field, string? value, int maxLength, string label)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            errors[field] = [$"{label} je obavezno."];
            return;
        }

        CheckLength(errors, field, value, maxLength, label);
    }

    /// <summary>Optional free text: only the length matters.</summary>
    public static void OptionalText(
        IDictionary<string, string[]> errors, string field, string? value, int maxLength, string label)
    {
        if (!string.IsNullOrWhiteSpace(value))
        {
            CheckLength(errors, field, value, maxLength, label);
        }
    }

    public static void RequirePassword(IDictionary<string, string[]> errors, string field, string? value)
    {
        if (string.IsNullOrWhiteSpace(value) || value.Length < MinPasswordLength)
        {
            errors[field] = [$"Lozinka mora imati najmanje {MinPasswordLength} karaktera."];
        }
    }

    private static void CheckLength(
        IDictionary<string, string[]> errors, string field, string value, int maxLength, string label)
    {
        // Trimmed, because every caller trims before persisting - validating the
        // untrimmed value would reject input that would actually have fit.
        if (value.Trim().Length > maxLength)
        {
            errors[field] = [$"{label} može imati najviše {maxLength} karaktera."];
        }
    }

    [GeneratedRegex(@"^[^@\s]+@[^@\s]+\.[^@\s]+$")]
    private static partial Regex EmailPattern();

    [GeneratedRegex(@"^\+?[0-9 ]{6,20}$")]
    private static partial Regex PhonePattern();
}
